src/compas/datastructures/assembly/part.py

```diff
@@ -0,0 +1,168 @@
+        'intersection': boolean_intersection_mesh_mesh
+    }
+
+    @property
+    def DATASCHEMA(self):
+        import schema
+        return schema.Schema({
+            "attributes": dict,
+            "key": int,
+            "frame": Frame,
+            "transformations": list,
+            "geometry": Shape,
+            "features": list
+        })
+
+    @property
+    def JSONSCHEMANAME(self):
+        return 'assembly'
+
+    def __init__(self, name, **kwargs):
+        super(Part, self).__init__()
+        self._key = None
+        self._frame = None
+        self._transformations = deque()
+        self._geometry = None
+        self._features = []
+        self._shape = None
+
+    def __str__(self):
+        tpl = "<Part with base geometry {} and features {}>"
+        return tpl.format(self.geometry, self.features)
+
+    @property
+    def data(self):
+        """dict : A data dict representing the part attributes, internal data structure, and geometries for serialization.
+        """
+        data = {
+            'attributes': self.attributes,
+            'mesh': self.mesh.data,
+            'graph': self.graph.data
+        }
+        return data
+
+    @data.setter
+    def data(self, data):
+        self.attributes.update(data['attributes'] or {})
+        self.mesh.data = data['mesh']
+        self.graph.data = data['graph']
+
+    @property
+    def key(self):
+        return self._key
+
+    @property
+    def frame(self):
+        if not self._frame:
+            self._frame = Frame.worldXY()
+        return self._frame
+
+    @frame.setter
+    def frame(self, frame):
+        if not isinstance(frame, Frame):
+            raise FrameError
+        self._frame = frame
+
+    @property
+    def transformations(self):
+        return self._transformations
+
+    @property
+    def geometry(self):
+        if not self._geometry:
+            self._geometry = Shape()
+        return self._geometry
+
+    @geometry.setter
+    def geometry(self, geometry):
+        self._geometry = geometry
+
+    @property
+    def features(self):
+        return self._features
+
+    @property
+    def shape(self):
+        return self._shape
+
+    def add_transformation(self, matrix):
+        """Add a transformation to the stack.
+
+        Parameters
+        ----------
+        matrix : :class:`compas.geometry.Transformation`
+        """
+        self._transformations.appendleft(matrix)
+
+    def add_feature(self, shape, operation):
+        """Add a feature to the geometry and the operation through which they should be combined.
+
+        Parameters
+        ----------
+        shape : :class:`compas.geometry.Shape`
+            The shape of the feature.
+        operation : {'union', 'difference', 'intersection'}
+            The boolean operation through which the feature should be integrated in the geometry.
+        """
+        if operation not in ('union', 'difference', 'intersection'):
```

Comment:
Maybe change to `if operation not in Part.operations.keys()`?

Reply:
done